src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
   assert(!extents.empty());
   DEBUGT("start with {} allocated extents",
          t, extents.size());
-  return trans_intr::do_for_each(extents,
-    [this, &t, FNAME](auto& ex) {
-    auto paddr = ex->get_paddr();
-    assert(paddr.is_absolute());
-    RandomBlockManager * rbm = rb_cleaner->get_rbm(paddr); 
-    assert(rbm);
-    TRACE("extent {}, allocated addr {}", fmt::ptr(ex.get()), paddr);
-    auto& stats = t.get_ool_write_stats();
-    stats.extents.num += 1;
-    stats.extents.bytes += ex->get_length();
-    stats.num_records += 1;
-
-    ex->prepare_write();
-    extent_len_t offset = 0;
-    bufferptr bp;
-    if (can_inplace_rewrite(t, ex)) {
-      assert(ex->is_logical());
-      auto r = ex->template cast<LogicalCachedExtent>()->get_modified_region();
-      ceph_assert(r.has_value());
-      offset = p2align(r->offset, rbm->get_block_size());
-      extent_len_t len =
-	p2roundup(r->offset + r->len, rbm->get_block_size()) - offset;
-      bp = ceph::bufferptr(ex->get_bptr(), offset, len);
-    } else {
-      bp = ex->get_bptr();
-      auto& trans_stats = get_by_src(w_stats.stats_by_src, t.get_src());
-      ++(trans_stats.num_records);
-      trans_stats.data_bytes += ex->get_length();
-      w_stats.data_bytes += ex->get_length();
-    }
-    return trans_intr::make_interruptible(
-      rbm->write(paddr + offset,
-	bp
-      ).handle_error(
-	alloc_write_iertr::pass_further{},
-	crimson::ct_error::assert_all{
-	  "Invalid error when writing record"}
-      )
-    ).si_then([this, &t, &ex, paddr, FNAME] {
-      TRACET("ool extent written at {} -- {}",
-	     t, paddr, *ex);
-      if (ex->is_initial_pending()) {
-	t.mark_allocated_extent_ool(ex);
-      } else if (can_inplace_rewrite(t, ex)) {
-        assert(ex->is_logical());
-	t.mark_inplace_rewrite_extent_ool(
-          ex->template cast<LogicalCachedExtent>());
+
+  struct io_desc_t {
+    paddr_t base;
+    extent_len_t len = 0;
+    std::list<CachedExtentRef> consecutive_extents;
+  };
+
+  return seastar::do_with(
+      std::vector<io_desc_t>(),
+      [this, &t, extents, FNAME](auto& io) {
+
+    for (auto& cur : extents) {
+      auto& stats = t.get_ool_write_stats();
+      stats.extents.num += 1;
+      stats.extents.bytes += cur->get_length();
+      stats.num_records += 1;
```

Comment:
Probably `num_records` can be increased once per batch, similar to the use case of the segmented backend.

Then we will be able to understand how merging is really working for RBM.